src/utils/src/Semaphore.c

```diff
@@ -88,7 +113,12 @@ STATUS semaphoreCreateInternal(UINT32 maxPermits, PSemaphore* ppSemaphore)
     CHK(maxPermits > 0, STATUS_INVALID_ARG);
     CHK(NULL != (pSemaphore = (PSemaphore) MEMCALLOC(1, SIZEOF(Semaphore))), STATUS_NOT_ENOUGH_MEMORY);
     pSemaphore->maxPermitCount = maxPermits;
-    ATOMIC_STORE(&pSemaphore->permitCount, maxPermits);
+    pSemaphore->signalSemaphore = empty;
+    if (empty) {
+        ATOMIC_STORE(&pSemaphore->permitCount, 0);
```

Comment:
If the permitCount is set to 0, no resource would actually be able to acquire the semaphore per se right? 

This is the flow I am seeing when a thread tries to acquire the semaphore:
```
semaphoreAcquireInternal(...) {
    existingCount = ATOMIC_DECREMENT(&pSemaphore->permitCount); ---> existingCount = 0, permitCount = -1
    if ((INT32) existingCount <= 0) { ---> this evaluates to TRUE
    acquireFailed = TRUE; 
 CleanUp: 
     if (acquireFailed) { ---> TRUE
        ATOMIC_INCREMENT(&pSemaphore->permitCount); ---> permitCount = 0 since it was decremented to -1 previously
    }
```

Which means semaphore acquire would never succeed. Is that the intention?

Reply:
It won't succeed until semaphoreRelease is called. It's mechanically the same as blocking/waiting for a signal.